File: python/src/prts_mcp/data/stage.py
from __future__ import annotations

import re as _re
from functools import lru_cache as _lru_cache

from prts_mcp.config import Config as _Config
from prts_mcp.data.stores import DirectoryStore as _DirectoryStore

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

_STAGE_FILE = "stage_table.json"
_ZONE_FILE = "zone_table.json"

_STAGE_TYPE_LABELS: dict[str, str] = {
    "MAIN": "主线",
    "ACTIVITY": "活动",
    "SUB": "支线",
    "DAILY": "每日",
    "CAMPAIGN": "剿灭",
    "CLIMB_TOWER": "爬塔",
    "SPECIAL_STORY": "特殊故事",
    "GUIDE": "教程",
}

_DIFFICULTY_LABELS: dict[str, str] = {
    "NORMAL": "普通",
    "FOUR_STAR": "突袭",
    "SIX_STAR": "六星",
}

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _get_config() -> _Config:
    return _Config.load()


def _store() -> _DirectoryStore:
    ep = _get_config().effective_excel_path
    if ep is None:
        raise RuntimeError("effective_excel_path is None — GAMEDATA_PATH may be unset")
    return _DirectoryStore(ep)


def _has_stage_data() -> bool:
    cfg = _get_config()
    if cfg.effective_excel_path is None:
        return False
    return _store().exists(_STAGE_FILE)


def _missing_data_message() -> str:
    return (
        "关卡数据暂不可用。请检查 GAMEDATA_PATH 配置，"
        "或等待服务器自动从 GitHub Release 同步数据完成后重试。"
    )


def _stage_type_label(t: str) -> str:
    return _STAGE_TYPE_LABELS.get(t, t)


def _difficulty_label(d: str) -> str:
    return _DIFFICULTY_LABELS.get(d, d)


def _clean_description(desc: str) -> str:
    """Strip angle-bracket markup like <@lv.fs> and </>."""
    if not desc:
        return ""
    return _re.sub(r"<[^>]+>", "", desc).strip()


def _format_unlock(conditions: list) -> str:
    if not conditions:
        return "（无条件）"
    state_labels = {
        "PASS": lambda sid: f"通关 {sid}",
        "STAR_3": lambda sid: f"三星通关 {sid}",
    }
    parts: list[str] = []
    for c in conditions:
        sid = c.get("stageId", "?")
        cs = c.get("completeState", "")
        parts.append(state_labels.get(cs, lambda s: f"{cs} {s}")(sid))
    return "；".join(parts)


def _format_drops(drop_info: dict | None) -> str:
    if not drop_info:
        return "（无）"
    display = drop_info.get("displayRewards") or []
    if not display:
        return "（无）"
    parts: list[str] = []
    for d in display:
        name = d.get("type") or d.get("dropType") or "?"
        count = d.get("count", 1)
        parts.append(f"- {name} ×{count}")
    return "\n".join(parts) if parts else "（无）"


# ---------------------------------------------------------------------------
# Cached loaders
# ---------------------------------------------------------------------------


@_lru_cache(maxsize=1)
def _load_stage_table() -> dict[str, dict]:
    if not _has_stage_data():
        raise FileNotFoundError(_STAGE_FILE)
    raw = _store().read_json(_STAGE_FILE)
    if not isinstance(raw, dict):
        raise TypeError(f"{_STAGE_FILE} top-level shape mismatch")
    stages = raw.get("stages")
    if not isinstance(stages, dict):
        raise TypeError(f"{_STAGE_FILE} missing 'stages' dict")
    return stages


@_lru_cache(maxsize=1)
def _load_zone_table() -> dict[str, dict] | None:
    store = _store()
    if not store.exists(_ZONE_FILE):
        return None
    raw = store.read_json(_ZONE_FILE)
    if not isinstance(raw, dict):
        return None
    zones = raw.get("zones")
    return zones if isinstance(zones, dict) else None


def _zone_display(zone_id: str) -> str:
    zones = _load_zone_table()
    if zones is None:
        return zone_id
    z = zones.get(zone_id)
    if z is None:
        return zone_id
    first = z.get("zoneNameFirst") or ""
    second = z.get("zoneNameSecond") or ""
    if first and second:
        return f"{first}-{second}"
    if first:
        return first
    return zone_id


def clear_stage_caches() -> None:
    _load_stage_table.cache_clear()
    _load_zone_table.cache_clear()


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def list_stages(
    chapter: str | None = None,
    type: str | None = None,
    limit: int = 50,
    offset: int = 0,
) -> str:
    """List stages, optionally filtered by zone ID and/or stage type."""
    if limit < 1:
        return "limit 必须 >= 1。"
    if offset < 0:
        return "offset 必须 >= 0。"
    if type is not None and type.upper() not in _STAGE_TYPE_LABELS:
        allowed = "、".join(_STAGE_TYPE_LABELS)
        return f"无效的 type：{type!r}。可选值：{allowed}。"

    try:
        stages = _load_stage_table()
    except (FileNotFoundError, TypeError) as e:
        return _missing_data_message() + f"（{e}）"

    filtered: list[dict] = []
    for sid, entry in sorted(stages.items()):
        if chapter is not None and entry.get("zoneId") != chapter:
            continue
        if type is not None and entry.get("stageType") != type.upper():
            continue
        filtered.append(entry)

    total = len(filtered)
    page = filtered[offset : offset + limit]

    if not page:
        if total == 0:
            filters: list[str] = []
            if chapter:
                filters.append(f"zoneId={chapter}")
            if type:
                filters.append(f"stageType={type.upper()}")
            return f"没有匹配的关卡（filter: {', '.join(filters) or 'none'}）。"
        return f"offset {offset} 超出范围（共 {total} 条）。"

    lines = [f"# 关卡列表（共 {total} 个）"]
    for e in page:
        t_label = _stage_type_label(e.get("stageType", ""))
        d_label = _difficulty_label(e.get("difficulty", ""))
        zd = _zone_display(e.get("zoneId", ""))
        name = e.get("name") or "（无名）"
        code = e.get("code") or "?"
        lines.append(
            f"- **{name}** [{t_label}] {code} — {d_label} — {zd}"
        )

    start = offset + 1
    end = min(offset + limit, total)
    lines.append(
        f"\n（显示第 {start}–{end} 条，共 {total} 条。"
        f"使用 offset={offset + limit} 查看下一页）"
    )
    return "\n".join(lines)


def get_stage_info(stage_id: str) -> str:
    """Return detailed information for a single stage."""
    try:
        stages = _load_stage_table()
    except (FileNotFoundError, TypeError) as e:
        return _missing_data_message() + f"（{e}）"

    entry: dict | None = stages.get(stage_id)
    if entry is None:
        return f"未找到关卡：{stage_id!r}。"

    name = entry.get("name") or "（无名）"
    code = entry.get("code") or "?"
    t_label = _stage_type_label(entry.get("stageType", ""))
    d_label = _difficulty_label(entry.get("difficulty", ""))
    zone_id = entry.get("zoneId", "")
    zd = _zone_display(zone_id)
    _ap = entry.get("apCost")
    ap = _ap if _ap is not None else "?"
    danger = entry.get("dangerLevel") or "?"
    boss = entry.get("bossMark", False)
    raw_desc = entry.get("description") or ""
    desc = _clean_description(raw_desc) or "（无描述）"
    drop_info = entry.get("stageDropInfo")
    unlocks = entry.get("unlockCondition") or []
    hard_id = entry.get("hardStagedId")
    level_id = entry.get("levelId")

    parts = [f"# {name} — 关卡详情", "", "## 基本信息"]
    parts.append(f"- **ID**：{stage_id}")
    parts.append(f"- **编号**：{code}")
    parts.append(f"- **类型**：{t_label}")
    parts.append(f"- **难度**：{d_label}")
    parts.append(f"- **所属区域**：{zd}")
    parts.append(f"- **理智消耗**：{ap}")
    parts.append(f"- **危险等级**：{danger}")
    if boss:
        parts.append("- **BOSS标记**：是")
    if level_id:
        parts.append(f"- **关卡数据**：{level_id}")

    parts.append("")
    parts.append("## 描述")
    parts.append(desc)

    parts.append("")
    parts.append("## 掉落信息")
    parts.append(_format_drops(drop_info))

    parts.append("")
    parts.append("## 解锁条件")
    parts.append(_format_unlock(unlocks))

    parts.append("")
    parts.append("## 关联关卡")
    if hard_id:
        h_entry = stages.get(hard_id)
        h_name = h_entry.get("name") if h_entry else None
        parts.append(f"- 突袭模式：{hard_id}" + (f"（{h_name}）" if h_name else ""))
    else:
        parts.append("- 突袭模式：无")
    ssid = entry.get("sixStarStageId")
    if ssid:
        s_entry = stages.get(ssid)
        s_name = s_entry.get("name") if s_entry else None
        parts.append(f"- 六星模式：{ssid}" + (f"（{s_name}）" if s_name else ""))

    return "\n".join(parts)


def search_stages(pattern: str, max_results: int = 30) -> str:
    """Regex search across stage names, codes, and descriptions."""
    if max_results < 1:
        return "max_results 必须 >= 1。"

    try:
        regex = _re.compile(pattern, _re.IGNORECASE)
    except _re.error as e:
        return f"正则表达式无效：{e}"

    try:
        stages = _load_stage_table()
    except (FileNotFoundError, TypeError) as e:
        return _missing_data_message() + f"（{e}）"

    matched: list[dict] = []
    for sid, entry in sorted(stages.items()):
        search_text = (
            (entry.get("name") or "")
            + " "
            + (entry.get("code") or "")
            + " "
            + _clean_description(entry.get("description") or "")
            + " "
            + (entry.get("stageType") or "")
        )
        if regex.search(search_text):
            matched.append(entry)
            if len(matched) >= max_results:
                break

    if not matched:
        return f"未找到匹配 '{pattern}' 的关卡。"

    lines = [f"# 搜索结果：{pattern}（共 {len(matched)} 个）"]
    for e in matched:
        name = e.get("name") or "（无名）"
        code = e.get("code") or "?"
        t_label = _stage_type_label(e.get("stageType", ""))
        d_label = _difficulty_label(e.get("difficulty", ""))
        zd = _zone_display(e.get("zoneId", ""))
        _ap = e.get("apCost")
        ap = _ap if _ap is not None else "?"
        raw_desc = e.get("description") or ""
        cdesc = _clean_description(raw_desc)

        lines.append(f"\n## {name} [{t_label}] {code}")
        lines.append(f"- **区域**：{zd}")
        lines.append(f"- **难度**：{d_label}")
        lines.append(f"- **理智**：{ap}")
        if cdesc:
            lines.append(f"- **描述**：{cdesc[:120]}{'...' if len(cdesc) > 120 else ''}")

    return "\n".join(lines)

File: ts/src/data/stage.ts
import { loadConfig } from "../config.js";
import { DirectoryStore } from "./stores.js";
import { getItemNameById } from "./item.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface StageEntry {
  stageId?: string;
  code?: string | null;
  name?: string | null;
  stageType?: string | null;
  difficulty?: string | null;
  zoneId?: string | null;
  levelId?: string | null;
  apCost?: number | null;
  dangerLevel?: string | null;
  description?: string | null;
  stageDropInfo?: Record<string, unknown> | null;
  unlockCondition?: { stageId: string; completeState: string }[] | null;
  hardStagedId?: string | null;
  sixStarStageId?: string | null;
  bossMark?: boolean | null;
}

type StageTable = Record<string, StageEntry>;

interface ZoneEntry {
  zoneID?: string;
  zoneNameFirst?: string | null;
  zoneNameSecond?: string | null;
}

type ZoneTable = Record<string, ZoneEntry>;

interface StageSearchRecord {
  stageId: string;
  entry: StageEntry;
  searchText: string;
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const STAGE_FILE = "stage_table.json";
const ZONE_FILE = "zone_table.json";

const STAGE_TYPE_LABELS: Record<string, string> = {
  MAIN: "主线",
  ACTIVITY: "活动",
  SUB: "支线",
  DAILY: "每日",
  CAMPAIGN: "剿灭",
  CLIMB_TOWER: "爬塔",
  SPECIAL_STORY: "特殊故事",
  GUIDE: "教程",
};

const DIFFICULTY_LABELS: Record<string, string> = {
  NORMAL: "普通",
  FOUR_STAR: "突袭",
  SIX_STAR: "六星",
};

// ---------------------------------------------------------------------------
// Module-level caches
// ---------------------------------------------------------------------------

let _stageTable: StageTable | null = null;
let _zoneTable: ZoneTable | null = null;
let _zoneTableFailed = false;
let _stageSearchRecords: StageSearchRecord[] | null = null;

export function clearStageCaches(): void {
  _stageTable = null;
  _zoneTable = null;
  _zoneTableFailed = false;
  _stageSearchRecords = null;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function stageTypeLabel(t: string): string {
  return STAGE_TYPE_LABELS[t] ?? t;
}

function difficultyLabel(d: string): string {
  return DIFFICULTY_LABELS[d] ?? d;
}

function cleanDescription(desc: string): string {
  if (!desc) return "";
  return desc.replace(/<[^>]+>/g, "").trim();
}

function formatUnlock(conditions: { stageId: string; completeState: string }[] | null): string {
  if (!conditions || conditions.length === 0) return "（无条件）";
  const labels: Record<string, (sid: string) => string> = {
    PASS: (sid) => `通关 ${sid}`,
    STAR_3: (sid) => `三星通关 ${sid}`,
  };
  const parts = conditions.map((c) => {
    const fn = labels[c.completeState] ?? ((s: string) => `${c.completeState} ${s}`);
    return fn(c.stageId);
  });
  return parts.join("；");
}

function formatDrops(dropInfo: Record<string, unknown> | null | undefined): string {
  if (!dropInfo) return "（无）";
  const display = (dropInfo["displayRewards"] ?? []) as Record<string, unknown>[];
  if (!Array.isArray(display) || display.length === 0) return "（无）";
  const parts = display.map((d) => {
    const itemId = String(d["id"] ?? "");
    const itemName = itemId ? getItemNameById(itemId) : null;
    let name = String(itemName ?? d["type"] ?? d["dropType"] ?? itemId ?? "?");
    if (itemId && itemName) name = `${name}（${itemId}）`;
    else if (itemId && name !== itemId) name = `${name}（${itemId}）`;
    const count = (d["count"] as number) ?? 1;
    const dropType = d["dropType"] ? ` [${String(d["dropType"])}]` : "";
    return `- ${name} ×${count}${dropType}`;
  });
  return parts.length > 0 ? parts.join("\n") : "（无）";
}

// ---------------------------------------------------------------------------
// Lazy loaders
// ---------------------------------------------------------------------------

function getStageTable(): StageTable {
  if (_stageTable === null) {
    const cfg = loadConfig();
    if (!cfg.effectiveExcelPath) {
      throw new Error("关卡数据暂不可用。请检查 GAMEDATA_PATH 配置。");
    }
    const store = new DirectoryStore(cfg.effectiveExcelPath);
    if (!store.exists(STAGE_FILE)) {
      throw new Error(`关卡数据文件不存在：${STAGE_FILE}`);
    }
    const raw = store.readJson<{ stages?: StageTable }>(STAGE_FILE);
    if (!raw || typeof raw !== "object" || !raw.stages || typeof raw.stages !== "object") {
      throw new Error(`${STAGE_FILE} 格式异常`);
    }
    _stageTable = raw.stages;
  }
  return _stageTable;
}

function getZoneTable(): ZoneTable | null {
  if (_zoneTable === null && !_zoneTableFailed) {
    const cfg = loadConfig();
    if (!cfg.effectiveExcelPath) {
      _zoneTableFailed = true;
      return null;
    }
    const store = new DirectoryStore(cfg.effectiveExcelPath);
    if (!store.exists(ZONE_FILE)) {
      _zoneTableFailed = true;
      return null;
    }
    const raw = store.readJson<{ zones?: ZoneTable }>(ZONE_FILE);
    if (!raw || typeof raw !== "object" || !raw.zones || typeof raw.zones !== "object") {
      _zoneTableFailed = true;
      return null;
    }
    _zoneTable = raw.zones;
  }
  return _zoneTable;
}

function zoneDisplay(zoneId: string): string {
  const zones = getZoneTable();
  if (!zones) return zoneId;
  const z = zones[zoneId];
  if (!z) return zoneId;
  const first = z.zoneNameFirst || "";
  const second = z.zoneNameSecond || "";
  if (first && second) return `${first}-${second}`;
  if (first) return first;
  return zoneId;
}

function missingDataMessage(): string {
  return (
    "关卡数据暂不可用。请检查 GAMEDATA_PATH 配置，" +
    "或等待服务器自动从 GitHub Release 同步数据完成后重试。"
  );
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export function listStages(
  chapter?: string | null,
  type?: string | null,
  limit: number = 50,
  offset: number = 0,
): string {
  if (limit < 1) return "limit 必须 >= 1。";
  if (offset < 0) return "offset 必须 >= 0。";
  if (type != null && !(type.toUpperCase() in STAGE_TYPE_LABELS)) {
    const allowed = Object.keys(STAGE_TYPE_LABELS).join("、");
    return `无效的 type：${JSON.stringify(type)}。可选值：${allowed}。`;
  }

  let stages: StageTable;
  try {
    stages = getStageTable();
  } catch (e) {
    return missingDataMessage() + `（${e instanceof Error ? e.message : String(e)}）`;
  }

  const filtered: StageEntry[] = [];
  for (const [sid, entry] of Object.entries(stages).sort(([a], [b]) => a.localeCompare(b))) {
    if (chapter != null && entry.zoneId !== chapter) continue;
    if (type != null && entry.stageType !== type.toUpperCase()) continue;
    filtered.push(entry);
  }

  const total = filtered.length;
  const page = filtered.slice(offset, offset + limit);

  if (page.length === 0) {
    if (total === 0) {
      const filters: string[] = [];
      if (chapter) filters.push(`zoneId=${chapter}`);
      if (type) filters.push(`stageType=${type.toUpperCase()}`);
      return `没有匹配的关卡（filter: ${filters.join(", ") || "none"}）。`;
    }
    return `offset ${offset} 超出范围（共 ${total} 条）。`;
  }

  const lines = [`# 关卡列表（共 ${total} 个）`];
  for (const e of page) {
    const tLabel = stageTypeLabel(e.stageType ?? "");
    const dLabel = difficultyLabel(e.difficulty ?? "");
    const zd = zoneDisplay(e.zoneId ?? "");
    const name = e.name || "（无名）";
    const code = e.code || "?";
    const sid = e.stageId ?? "";
    lines.push(`- **${name}** [${tLabel}] ${code} — ${dLabel} — ${zd}（id: ${sid}）`);
  }

  const start = offset + 1;
  const end = Math.min(offset + limit, total);
  lines.push(
    `\n（显示第 ${start}–${end} 条，共 ${total} 条。` +
    `使用 offset=${offset + limit} 查看下一页）`,
  );
  return lines.join("\n");
}

export function getStageInfo(stageId: string): string {
  let stages: StageTable;
  try {
    stages = getStageTable();
  } catch (e) {
    return missingDataMessage() + `（${e instanceof Error ? e.message : String(e)}）`;
  }

  const entry = stages[stageId];
  if (!entry) return `未找到关卡：${JSON.stringify(stageId)}。`;

  const name = entry.name || "（无名）";
  const code = entry.code || "?";
  const tLabel = stageTypeLabel(entry.stageType ?? "");
  const dLabel = difficultyLabel(entry.difficulty ?? "");
  const zd = zoneDisplay(entry.zoneId ?? "");
  const ap = entry.apCost ?? "?";
  const danger = entry.dangerLevel || "?";
  const boss = entry.bossMark === true;
  const rawDesc = entry.description || "";
  const desc = cleanDescription(rawDesc) || "（无描述）";
  const drops = entry.stageDropInfo as Record<string, unknown> | null | undefined;
  const unlocks = entry.unlockCondition ?? [];
  const hardId = entry.hardStagedId;
  const levelId = entry.levelId;

  const parts: string[] = [`# ${name} — 关卡详情`, "", "## 基本信息"];
  parts.push(`- **ID**：${stageId}`);
  parts.push(`- **编号**：${code}`);
  parts.push(`- **类型**：${tLabel}`);
  parts.push(`- **难度**：${dLabel}`);
  parts.push(`- **所属区域**：${zd}`);
  parts.push(`- **理智消耗**：${ap}`);
  parts.push(`- **危险等级**：${danger}`);
  if (boss) parts.push("- **BOSS标记**：是");
  if (levelId) parts.push(`- **关卡数据**：${levelId}`);

  parts.push("", "## 描述", desc);
  parts.push("", "## 掉落信息", formatDrops(drops));
  parts.push("", "## 解锁条件", formatUnlock(unlocks));

  parts.push("", "## 关联关卡");
  if (hardId) {
    const hEntry = stages[hardId];
    const hName = hEntry?.name;
    parts.push(`- 突袭模式：${hardId}` + (hName ? `（${hName}）` : ""));
  } else {
    parts.push("- 突袭模式：无");
  }
  const ssid = entry.sixStarStageId;
  if (ssid) {
    const sEntry = stages[ssid];
    const sName = sEntry?.name;
    parts.push(`- 六星模式：${ssid}` + (sName ? `（${sName}）` : ""));
  }

  return parts.join("\n");
}

export function searchStages(pattern: string, maxResults: number = 30): string {
  if (maxResults < 1) return "max_results 必须 >= 1。";
  if (maxResults > 100) return "max_results 必须 <= 100。";

  let regex: RegExp;
  try {
    regex = new RegExp(pattern, "i");
  } catch (e) {
    return `正则表达式无效：${e instanceof Error ? e.message : String(e)}`;
  }

  let records: StageSearchRecord[];
  try {
    records = getStageSearchRecords();
  } catch (e) {
    return missingDataMessage() + `（${e instanceof Error ? e.message : String(e)}）`;
  }

  const matched: StageSearchRecord[] = [];
  for (const record of records) {
    if (regex.test(record.searchText)) {
      matched.push(record);
      if (matched.length >= maxResults) break;
    }
  }

  if (matched.length === 0) return `未找到匹配 '${pattern}' 的关卡。`;

  const lines = [`# 搜索结果：${pattern}（共 ${matched.length} 个）`];
  for (const record of matched) {
    const e = record.entry;
    const name = e.name || "（无名）";
    const code = e.code || "?";
    const tLabel = stageTypeLabel(e.stageType ?? "");
    const dLabel = difficultyLabel(e.difficulty ?? "");
    const zd = zoneDisplay(e.zoneId ?? "");
    const ap = e.apCost ?? "?";
    const cdesc = cleanDescription(e.description ?? "");

    const sid = record.stageId;
    lines.push(`\n## ${name} [${tLabel}] ${code}（id: ${sid}）`);
    lines.push(`- **区域**：${zd}`);
    lines.push(`- **难度**：${dLabel}`);
    lines.push(`- **理智**：${ap}`);
    if (cdesc) lines.push(`- **描述**：${cdesc.slice(0, 120)}${cdesc.length > 120 ? "..." : ""}`);
  }

  return lines.join("\n");
}

function getStageSearchRecords(): StageSearchRecord[] {
  if (_stageSearchRecords !== null) return _stageSearchRecords;
  _stageSearchRecords = Object.entries(getStageTable())
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([stageId, entry]) => ({
      stageId,
      entry,
      searchText: [
        entry.name ?? "",
        entry.code ?? "",
        cleanDescription(entry.description ?? ""),
        entry.stageType ?? "",
        stageId,
      ].join(" "),
    }));
  return _stageSearchRecords;
}
